File: apps/desktop/electron/db/postgres.ts
import { Client } from 'pg'
import { DatabaseConnection, ConnectionConfig, StorageAnalysis, QueryResult, SchemaInfo } from './index'
import { copyFileSync, existsSync } from 'fs'
import { join } from 'path'

export class PostgreSQLConnection extends DatabaseConnection {
  private client?: Client

  async connect(): Promise<void> {
    this.client = new Client({
      host: this.config.host,
      port: this.config.port || 5432,
      database: this.config.database,
      user: this.config.username,
      password: this.config.password,
      connectionString: this.config.connectionString,
    })

    await this.client.connect()
    this.connected = true
  }

  async disconnect(): Promise<void> {
    if (this.client) {
      await this.client.end()
      this.client = undefined
      this.connected = false
    }
  }

  async testConnection(): Promise<void> {
    if (!this.client) {
      await this.connect()
    }
    
    const result = await this.client!.query('SELECT 1')
    if (result.rows.length === 0) {
      throw new Error('Connection test failed')
    }
  }

  async analyzeStorage(): Promise<StorageAnalysis> {
    if (!this.client) {
      await this.connect()
    }

    const client = this.client!

    // Get database size
    const dbSizeResult = await client.query(`
      SELECT pg_size_pretty(pg_database_size(current_database())) as size,
             pg_database_size(current_database()) as size_bytes
    `)
    const totalSize = parseInt(dbSizeResult.rows[0].size_bytes)

    // Get table information
    const tablesResult = await client.query(`
      SELECT 
        schemaname,
        tablename,
        pg_size_pretty(pg_total_relation_size(schemaname||'.'||tablename)) as size,
        pg_total_relation_size(schemaname||'.'||tablename) as size_bytes,
        pg_size_pretty(pg_relation_size(schemaname||'.'||tablename)) as table_size,
        pg_relation_size(schemaname||'.'||tablename) as table_size_bytes,
        n_tup_ins + n_tup_upd + n_tup_del as row_count
      FROM pg_stat_user_tables 
      ORDER BY pg_total_relation_size(schemaname||'.'||tablename) DESC
    `)

    // Get index information
    const indexesResult = await client.query(`
      SELECT 
        schemaname,
        indexname,
        tablename,
        pg_size_pretty(pg_relation_size(schemaname||'.'||indexname)) as size,
        pg_relation_size(schemaname||'.'||indexname) as size_bytes
      FROM pg_stat_user_indexes
      ORDER BY pg_relation_size(schemaname||'.'||indexname) DESC
    `)

    const tables = tablesResult.rows.map(row => ({
      name: `${row.schemaname}.${row.tablename}`,
      size: parseInt(row.size_bytes),
      rowCount: parseInt(row.row_count) || 0,
      indexSize: 0, // Will be calculated separately
      bloat: 0 // Placeholder for bloat calculation
    }))

    const indexes = indexesResult.rows.map(row => ({
      name: `${row.schemaname}.${row.indexname}`,
      tableName: `${row.schemaname}.${row.tablename}`,
      size: parseInt(row.size_bytes),
      bloat: 0 // Placeholder for bloat calculation
    }))

    // Calculate index sizes per table
    tables.forEach(table => {
      const tableIndexes = indexes.filter(idx => idx.tableName === table.name)
      table.indexSize = tableIndexes.reduce((sum, idx) => sum + idx.size, 0)
    })

    const largestTable = tables.length > 0 ? tables[0] : { name: 'N/A', size: 0, rowCount: 0 }

    return {
      totalSize,
      tableCount: tables.length,
      indexCount: indexes.length,
      largestTable,
      tables,
      indexes,
      lastAnalyzed: new Date().toISOString()
    }
  }

  async executeQuery(query: string): Promise<QueryResult> {
    if (!this.client) {
      await this.connect()
    }

    const startTime = Date.now()
    const result = await this.client!.query(query)
    const executionTime = Date.now() - startTime

    // Get explain plan if it's a SELECT query
    let explainPlan
    if (query.trim().toLowerCase().startsWith('select')) {
      try {
        const explainResult = await this.client!.query(`EXPLAIN (FORMAT JSON) ${query}`)
        explainPlan = explainResult.rows[0]['QUERY PLAN']
      } catch (error) {
        // Ignore explain errors
      }
    }

    return {
      columns: result.fields.map(field => field.name),
      rows: result.rows,
      rowCount: result.rowCount,
      executionTime,
      explainPlan
    }
  }

  async getSchema(): Promise<SchemaInfo> {
    if (!this.client) {
      await this.connect()
    }

    const client = this.client!

    // Get schemas
    const schemasResult = await client.query(`
      SELECT schema_name 
      FROM information_schema.schemata 
      WHERE schema_name NOT IN ('information_schema', 'pg_catalog', 'pg_toast')
      ORDER BY schema_name
    `)

    const schemas = []

    for (const schemaRow of schemasResult.rows) {
      const schemaName = schemaRow.schema_name

      // Get tables in this schema
      const tablesResult = await client.query(`
        SELECT table_name 
        FROM information_schema.tables 
        WHERE table_schema = $1 AND table_type = 'BASE TABLE'
        ORDER BY table_name
      `, [schemaName])

      const tables = []

      for (const tableRow of tablesResult.rows) {
        const tableName = tableRow.table_name

        // Get columns
        const columnsResult = await client.query(`
          SELECT 
            column_name,
            data_type,
            is_nullable,
            column_default
          FROM information_schema.columns 
          WHERE table_schema = $1 AND table_name = $2
          ORDER BY ordinal_position
        `, [schemaName, tableName])

        const columns = columnsResult.rows.map(col => ({
          name: col.column_name,
          type: col.data_type,
          nullable: col.is_nullable === 'YES',
          defaultValue: col.column_default
        }))

        // Get indexes
        const indexesResult = await client.query(`
          SELECT 
            indexname,
            indexdef
          FROM pg_indexes 
          WHERE schemaname = $1 AND tablename = $2
        `, [schemaName, tableName])

        const indexes = indexesResult.rows.map(idx => ({
          name: idx.indexname,
          columns: [], // Would need to parse indexdef to extract columns
          unique: idx.indexdef.includes('UNIQUE')
        }))

        tables.push({
          name: tableName,
          columns,
          indexes
        })
      }

      schemas.push({
        name: schemaName,
        tables
      })
    }

    return { schemas }
  }

  async createBackup(backupPath: string): Promise<{ path: string; size: number }> {
    if (!this.client) {
      await this.connect()
    }

    // For PostgreSQL, we'll use pg_dump via child_process
    const { spawn } = require('child_process')
    const { promisify } = require('util')
    const { writeFile } = require('fs').promises

    return new Promise((resolve, reject) => {
      const pgDump = spawn('pg_dump', [
        '--host', this.config.host || 'localhost',
        '--port', (this.config.port || 5432).toString(),
        '--username', this.config.username || '',
        '--dbname', this.config.database || '',
        '--format', 'custom',
        '--file', backupPath
      ])

      if (this.config.password) {
        pgDump.env.PGPASSWORD = this.config.password
      }

      pgDump.on('close', async (code: number) => {
        if (code === 0) {
          try {
            const stats = await require('fs').promises.stat(backupPath)
            resolve({ path: backupPath, size: stats.size })
          } catch (error) {
            reject(new Error('Failed to get backup file size'))
          }
        } else {
          reject(new Error(`pg_dump failed with code ${code}`))
        }
      })

      pgDump.on('error', (error: Error) => {
        reject(new Error(`pg_dump error: ${error.message}`))
      })
    })
  }

  async restoreBackup(backupPath: string): Promise<void> {
    if (!this.client) {
      await this.connect()
    }

    const { spawn } = require('child_process')

    return new Promise((resolve, reject) => {
      const pgRestore = spawn('pg_restore', [
        '--host', this.config.host || 'localhost',
        '--port', (this.config.port || 5432).toString(),
        '--username', this.config.username || '',
        '--dbname', this.config.database || '',
        '--clean',
        '--if-exists',
        backupPath
      ])

      if (this.config.password) {
        pgRestore.env.PGPASSWORD = this.config.password
      }

      pgRestore.on('close', (code: number) => {
        if (code === 0) {
          resolve()
        } else {
          reject(new Error(`pg_restore failed with code ${code}`))
        }
      })

      pgRestore.on('error', (error: Error) => {
        reject(new Error(`pg_restore error: ${error.message}`))
      })
    })
  }
}

File: db_storage_manager/db/postgres.py
"""
PostgreSQL database connection
"""

import asyncio
import subprocess
from pathlib import Path
from typing import Any, Dict, List
import psycopg2
from psycopg2.extras import RealDictCursor

from .base import (
    DatabaseConnection,
    ConnectionConfig,
    StorageAnalysis,
    QueryResult,
    SchemaInfo,
)


class PostgreSQLConnection(DatabaseConnection):
    """PostgreSQL database connection"""

    def __init__(self, config: ConnectionConfig):
        super().__init__(config)
        self.connection = None

    async def connect(self) -> None:
        """Connect to PostgreSQL database"""
        if self.config.connection_string:
            self.connection = psycopg2.connect(self.config.connection_string)
        else:
            self.connection = psycopg2.connect(
                host=self.config.host or "localhost",
                port=self.config.port or 5432,
                database=self.config.database,
                user=self.config.username,
                password=self.config.password,
            )
        self.connected = True

    async def disconnect(self) -> None:
        """Disconnect from PostgreSQL database"""
        if self.connection:
            self.connection.close()
            self.connection = None
        self.connected = False

    async def analyze_storage(self) -> StorageAnalysis:
        """Analyze PostgreSQL storage"""
        if not self.connection:
            await self.connect()

        cursor = self.connection.cursor(cursor_factory=RealDictCursor)

        # Get table sizes
        cursor.execute("""
            SELECT 
                schemaname || '.' || tablename as name,
                pg_total_relation_size(schemaname || '.' || tablename) as size,
                pg_relation_size(schemaname || '.' || tablename) as table_size,
                (SELECT COUNT(*) FROM information_schema.tables 
                 WHERE table_schema = t.schemaname AND table_name = t.tablename) as row_count
            FROM pg_tables t
            WHERE schemaname NOT IN ('pg_catalog', 'information_schema')
            ORDER BY size DESC
        """)

        tables = []
        total_size = 0
        for row in cursor.fetchall():
            # Get actual row count
            cursor.execute(f'SELECT COUNT(*) FROM "{row["name"].split(".")[0]}"."{row["name"].split(".")[1]}"')
            row_count = cursor.fetchone()[0]

            table_size = row["table_size"] or 0
            index_size = (row["size"] or 0) - table_size

            tables.append({
                "name": row["name"],
                "size": table_size,
                "rowCount": row_count,
                "indexSize": index_size,
                "bloat": 0.0,
            })
            total_size += row["size"] or 0

        # Get indexes
        cursor.execute("""
            SELECT 
                schemaname || '.' || indexname as name,
                tablename as table_name,
                pg_relation_size(indexrelid) as size
            FROM pg_indexes
            WHERE schemaname NOT IN ('pg_catalog', 'information_schema')
        """)

        indexes = []
        for row in cursor.fetchall():
            indexes.append({
                "name": row["name"],
                "tableName": row["table_name"],
                "size": row["size"] or 0,
                "bloat": 0.0,
            })

        largest_table = tables[0] if tables else {"name": "N/A", "size": 0, "rowCount": 0}

        return {
            "totalSize": total_size,
            "tableCount": len(tables),
            "indexCount": len(indexes),
            "largestTable": largest_table,
            "tables": tables,
            "indexes": indexes,
            "lastAnalyzed": __import__("datetime").datetime.now().isoformat(),
        }

    async def execute_query(self, query: str, safe_mode: bool = True) -> QueryResult:
        """Execute a PostgreSQL query"""
        if not self.connection:
            await self.connect()

        if safe_mode and not query.strip().upper().startswith("SELECT"):
            raise ValueError("Only SELECT queries are allowed in safe mode")

        cursor = self.connection.cursor(cursor_factory=RealDictCursor)
        import time
        start_time = time.time()

        try:
            cursor.execute(query)
            if cursor.description:
                columns = [desc[0] for desc in cursor.description]
                rows = [dict(row) for row in cursor.fetchall()]
            else:
                columns = ["affected_rows"]
                rows = [{"affected_rows": cursor.rowcount}]

            execution_time = int((time.time() - start_time) * 1000)

            # Get explain plan for SELECT queries
            explain_plan = None
            if query.strip().upper().startswith("SELECT"):
                try:
                    cursor.execute(f"EXPLAIN {query}")
                    explain_plan = [row[0] for row in cursor.fetchall()]
                except Exception:
                    pass

            return {
                "columns": columns,
                "rows": rows,
                "rowCount": len(rows),
                "executionTime": execution_time,
                "explainPlan": explain_plan,
            }
        finally:
            cursor.close()

    async def get_schema(self) -> SchemaInfo:
        """Get PostgreSQL schema"""
        if not self.connection:
            await self.connect()

        cursor = self.connection.cursor(cursor_factory=RealDictCursor)

        # Get schemas
        cursor.execute("""
            SELECT schema_name 
            FROM information_schema.schemata 
            WHERE schema_name NOT IN ('pg_catalog', 'information_schema', 'pg_toast')
        """)

        schemas = []
        for schema_row in cursor.fetchall():
            schema_name = schema_row["schema_name"]

            # Get tables
            cursor.execute("""
                SELECT table_name 
                FROM information_schema.tables 
                WHERE table_schema = %s
            """, (schema_name,))

            tables = []
            for table_row in cursor.fetchall():
                table_name = table_row["table_name"]

                # Get columns
                cursor.execute("""
                    SELECT 
                        column_name,
                        data_type,
                        is_nullable,
                        column_default
                    FROM information_schema.columns
                    WHERE table_schema = %s AND table_name = %s
                    ORDER BY ordinal_position
                """, (schema_name, table_name))

                columns = []
                for col in cursor.fetchall():
                    columns.append({
                        "name": col["column_name"],
                        "type": col["data_type"],
                        "nullable": col["is_nullable"] == "YES",
                        "defaultValue": col["column_default"],
                    })

                # Get indexes
                cursor.execute("""
                    SELECT 
                        indexname,
                        indexdef
                    FROM pg_indexes
                    WHERE schemaname = %s AND tablename = %s
                """, (schema_name, table_name))

                indexes = []
                for idx in cursor.fetchall():
                    indexes.append({
                        "name": idx["indexname"],
                        "columns": [],
                        "unique": "UNIQUE" in idx["indexdef"].upper(),
                    })

                tables.append({
                    "name": table_name,
                    "columns": columns,
                    "indexes": indexes,
                })

            schemas.append({
                "name": schema_name,
                "tables": tables,
            })

        return {"schemas": schemas}

    async def create_backup(self, backup_path: str) -> Dict[str, Any]:
        """Create PostgreSQL backup using pg_dump"""
        if not self.connection:
            await self.connect()

        backup_file = Path(backup_path)
        backup_file.parent.mkdir(parents=True, exist_ok=True)

        # Build pg_dump command
        cmd = [
            "pg_dump",
            "--host", self.config.host or "localhost",
            "--port", str(self.config.port or 5432),
            "--username", self.config.username,
            "--dbname", self.config.database,
            "--format", "c",
            "--file", str(backup_file),
        ]

        # Set password via environment variable
        env = __import__("os").environ.copy()
        if self.config.password:
            env["PGPASSWORD"] = self.config.password

        # Run pg_dump
        process = await asyncio.create_subprocess_exec(
            *cmd,
            env=env,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        stdout, stderr = await process.communicate()

        if process.returncode != 0:
            raise RuntimeError(f"pg_dump failed: {stderr.decode()}")

        size = backup_file.stat().st_size
        return {"path": str(backup_file), "size": size}

    async def restore_backup(self, backup_path: str) -> None:
        """Restore PostgreSQL backup using pg_restore"""
        if not self.connection:
            await self.connect()

        backup_file = Path(backup_path)
        if not backup_file.exists():
            raise FileNotFoundError(f"Backup file not found: {backup_path}")

        # Build pg_restore command
        cmd = [
            "pg_restore",
            "--host", self.config.host or "localhost",
            "--port", str(self.config.port or 5432),
            "--username", self.config.username,
            "--dbname", self.config.database,
            "--clean",
            "--if-exists",
            str(backup_file),
        ]

        # Set password via environment variable
        env = __import__("os").environ.copy()
        if self.config.password:
            env["PGPASSWORD"] = self.config.password

        # Run pg_restore
        process = await asyncio.create_subprocess_exec(
            *cmd,
            env=env,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        stdout, stderr = await process.communicate()

        if process.returncode != 0:
            raise RuntimeError(f"pg_restore failed: {stderr.decode()}")

